### PR TITLE
feat(logging): add a `tracing span` per JSON-RPC call

### PR DESCRIPTION
**Pull request by niklasad1**
Essentially this PR downgrades to only print request and response bodies when `trace!` is enabled.

In addition I have added a tracing span per call so it's possible to just trace specific method calls which could be quite useful I reckon.

For batches it's possible to just `log/trace` by the target `batch` because the span is called `batch` so not possible to trace the individual calls inside it.

The main motivation for tracing spans is for the servers where this can be super useful.

## Example WS

```bash
2022-05-27T16:02:00.688793Z DEBUG jsonrpsee_client_transport::ws: Connecting to target: Target { sockaddrs: [127.0.0.1:35045], host: "127.0.0.1", host_header: "127.0.0.1:35045", _mode: Plain, path_and_query: "/" }
2022-05-27T16:02:00.689003Z  INFO jsonrpsee_ws_server::server: Accepting new connection 1/100
2022-05-27T16:02:00.689562Z  INFO jsonrpsee_client_transport::ws: Connection established to target: Target { sockaddrs: [], host: "127.0.0.1", host_header: "127.0.0.1:35045", _mode: Plain, path_and_query: "/" }
2022-05-27T16:02:00.689759Z TRACE method_call{method=say_hello}: jsonrpsee_core::tracing: send="{\"jsonrpc\":\"2.0\",\"id\":0,\"method\":\"say_hello\"}"
2022-05-27T16:02:00.689826Z TRACE jsonrpsee_client_transport::ws: send: {"jsonrpc":"2.0","id":0,"method":"say_hello"}
2022-05-27T16:02:00.690105Z TRACE method_call{method=say_hello}: jsonrpsee_core::tracing: recv="{\"jsonrpc\":\"2.0\",\"id\":0,\"method\":\"say_hello\",\"params\":null}"
2022-05-27T16:02:00.690160Z TRACE method_call{method=say_hello}: jsonrpsee_core::tracing: send="{\"jsonrpc\":\"2.0\",\"result\":\"lo\",\"id\":0}"
2022-05-27T16:02:00.690394Z TRACE method_call{method=say_hello}: jsonrpsee_core::tracing: recv="{\"jsonrpc\":\"2.0\",\"result\":\"lo\",\"id\":0}"
```

## Example HTTP

```bash
2022-05-27T15:30:09.114372Z TRACE method_call{method=say_hello}: jsonrpsee_core::tracing: send="{\"jsonrpc\":\"2.0\",\"id\":0,\"method\":\"say_hello\",\"params\":[1,2,3]}"
2022-05-27T15:30:09.114518Z TRACE method_call{method=say_hello}: hyper::client::pool: checkout waiting for idle connection: ("http", 127.0.0.1:45057)
2022-05-27T15:30:09.114600Z TRACE method_call{method=say_hello}: hyper::client::connect::http: Http::connect; scheme=Some("http"), host=Some("127.0.0.1"), port=Some(Port(45057))
2022-05-27T15:30:09.114653Z DEBUG method_call{method=say_hello}: hyper::client::connect::http: connecting to 127.0.0.1:45057
2022-05-27T15:30:09.114761Z TRACE method_call{method=say_hello}: mio::poll: registering event source with poller: token=Token(2), interests=READABLE | WRITABLE
2022-05-27T15:30:09.114968Z DEBUG method_call{method=say_hello}: hyper::client::connect::http: connected to 127.0.0.1:45057
2022-05-27T15:30:09.115012Z TRACE method_call{method=say_hello}: hyper::client::conn: client handshake Http1
2022-05-27T15:30:09.115048Z TRACE method_call{method=say_hello}: hyper::client::client: handshake complete, spawning background dispatcher task
2022-05-27T15:30:09.115162Z TRACE method_call{method=say_hello}: want: poll_want: taker wants!
2022-05-27T15:30:09.115199Z TRACE method_call{method=say_hello}: hyper::client::pool: checkout dropped for ("http", 127.0.0.1:45057)
2022-05-27T15:30:09.115715Z TRACE method_call{method=say_hello}: jsonrpsee_core::tracing: recv="{\"jsonrpc\":\"2.0\",\"id\":0,\"method\":\"say_hello\",\"params\":[1,2,3]}"
2022-05-27T15:30:09.115776Z TRACE method_call{method=say_hello}: jsonrpsee_core::tracing: send="{\"jsonrpc\":\"2.0\",\"result\":\"lo\",\"id\":0}"
2022-05-27T15:30:09.116148Z TRACE method_call{method=say_hello}: jsonrpsee_core::tracing: recv="{\"id\":0,\"jsonrpc\":\"2.0\",\"result\":\"lo\"}"
```

### CODE REVIEW

File: core/src/tracing.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+pub fn rx_log_from_bytes(bytes: &[u8], max: u32) {
+	if tracing::enabled!(Level::TRACE) {
+		let res = serde_json::from_slice::<serde_json::Value>(bytes).unwrap_or_default();
+		rx_log_from_json(&res, max);
+	}
+}
+
+/// Find the next char boundary to truncate at.
+fn truncate_at_char_boundary(s: &str, max: u32) -> &str {
+	match s.char_indices().nth(max as usize) {
```
Comment (jsdw) [inline core/src/tracing.rs:80]: This is quite an expensive thing to do relatively speaking, but I guess we only do it for TRACE level logs which are only used in exceptional cases!

A quick optimisation might be to check `s.len() < max` first, and if that is true, there is def no need to truncate?

Comment (jsdw) [inline core/src/tracing.rs:80]: This is quite an expensive thing to do relatively speaking, but I guess we only do it for TRACE level logs which are only used in exceptional cases!

A quick optimisation might be to check `s.len() < max` first, and if that is true, there is def no need to truncate?

**Review by jsdw — APPROVED** [review]
Assuming you still want to get this merged, looks good to me, and the tracing stuff sounds very useful!

**Review by lexnv — APPROVED** [review]

File: core/src/client/async_client/mod.rs [hunk 2]
```diff
@@ -198,6 +212,10 @@ pub struct Client {
 	request_timeout: Duration,
 	/// Request ID manager.
 	id_manager: RequestIdManager,
+	/// Max length for logging for requests and responses.
+	///
+	/// Entires bigger than this limit will be truncated.
```
Comment (niklasad1) [inline core/src/client/async_client/mod.rs:217]:
```suggestion
	/// Entries bigger than this limit will be truncated.
```

File: core/Cargo.toml [hunk 3]
```diff
@@ -15,14 +15,16 @@ jsonrpsee-types = { path = "../types", version = "0.14.0" }
...
+tracing = "0.1"
 
 # optional deps
 arrayvec = { version = "0.7.1", optional = true }
 async-channel = { version = "1.6", optional = true }
 async-lock = { version = "2.4", optional = true }
 futures-util = { version = "0.3.14", default-features = false, optional = true }
 hyper = { version = "0.14.10", default-features = false, features = ["stream"], optional = true }
-tracing = { version = "0.1", optional = true }
+
```
Comment (niklasad1) [inline core/Cargo.toml:26]:
```suggestion
```

File: core/src/client/async_client/mod.rs [hunk 4]
```diff
@@ -428,42 +461,40 @@ async fn handle_backend_messages<S: TransportSenderT, R: TransportReceiverT>(
...
 			}
 		}
 		// Unparsable response
 		else {
-			tracing::debug!(
-					"[backend]: recv unparseable message: {:?}",
-					serde_json::from_slice::<serde_json::Value>(&raw)
-				);
-			return Err(Error::Custom("Unparsable response".into()));
+			let json = serde_json::from_slice::<serde_json::Value>(&raw);
```
Comment (niklasad1) [inline core/src/client/async_client/mod.rs:490]: the connection is killed here, we can propagate the entire response in the error then.

File: core/src/lib.rs [hunk 5]
```diff
@@ -55,6 +55,9 @@ cfg_client! {
 	pub mod client;
 }
 
+/// Shared tracing helpers to trace RPC calls.
+pub mod tracing;
+
```
Comment (niklasad1) [inline core/src/lib.rs:60]:
```suggestion
```

File: core/src/tracing.rs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+#[derive(Debug)]
+/// Wrapper over [`tracing::Span`] to trace individual method calls, notifications and similar.
+pub struct RpcTracing(tracing::Span);
+
+impl RpcTracing {
+	/// Create a `method_call` tracing target.
+	///
+	/// To enable this you need to call `RpcTracing::method_call("some_method").span().enable()`.
+	pub fn method_call(method: &str) -> Self {
+		Self(tracing::span!(tracing::Level::DEBUG, "method_call", %method))
```
Comment (longbowlu) [inline core/src/tracing.rs:13]: ^

**Pull request merged**